core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,397 @@
+  private type ResultingRef = Ref[(Declaration,  NormalExpressionTag)]
+  private type Env = (Map[Identifier, NormalExpressionTag], List[Option[Identifier]])
+  private type NormState[A] = State[Map[(PackageName, Identifier), ResultingRef], A]
+
+  private def norm(input: (Package.Inferred, SourceRef, Env)): NormState[ResultingRef] =
+    input match {
+      case ((pack, Right(expr), env)) =>
+        for {
+          expr <- normalizeExpr(expr, env, pack)
+        } yield Right(expr)
+      case (pack, Left((item, t)), env) => {
+        NameKind(pack, item).get match { // this get should never fail due to type checking
+          case NameKind.Let(_, _, expr) =>
+            for {
+              lookup <- State.inspect {
+                lets: Map[(PackageName, Identifier), ResultingRef] =>
+                  lets.get((pack.name, item))
+              }
+              res <- lookup match {
+                case Some(res) =>
+                  State.pure(res): State[
+                  Map[(PackageName, Identifier), ResultingRef],
+                  ResultingRef]
+                case None =>
+                  for {
+                    res <- norm((pack, Right(expr), env))
+                    _ <- State.modify {
+                      lets: Map[(PackageName, Identifier), ResultingRef] =>
+                        lets + ((pack.name, item) -> res)
+                    }
+                  } yield res
+
+              }
+            } yield Left((item, getTag(res)))
+                case NameKind.Constructor(cn, _, dt, _) => {
+                  val neTag: NormalExpressionTag = NormalExpressionTag(constructor(cn, dt), Set())
+                  State.pure(Left((item, (t, neTag))))
+                }
+                case NameKind.Import(from, orig) => {
+                  // we reset the environment in the other package
+                  for {
+                    imported <- norm((pm.toMap(from.unfix.name), Left((orig, t)), (Map.empty, Nil)))
+                    neTag = getTag(imported)._2
+                  } yield Left((item, (t, neTag)))
+                }
+                case NameKind.ExternalDef(pn, n, scheme) => {
+                  val neTag = NormalExpressionTag(NormalExpression.ExternalVar(pn, n), Set())
+                  State.pure(Left((item, (t, neTag))))
+                }
+        }
+        }
+    }
+
+  private def constructor(c: Constructor, dt: rankn.DefinedType[Any]): NormalExpression = {
+      val (enum, arity) = dt.constructors
+        .toList
+        .iterator
+        .zipWithIndex
+        .collectFirst { case ((ctor, params, resType), idx) if ctor == c => (idx, params.size) }
+        .get
+
+      def loop(params: Int, expr: NormalExpression): NormalExpression =
+        if (params == 0) expr
+        else loop(params - 1, NormalExpression.Lambda(expr))
+
+        loop(arity, NormalExpression.Struct(enum, ((arity - 1) to 0 by -1).map(NormalExpression.LambdaVar(_)).toList))
```

Comment:
the order of the indices looks backwards to me, but I'm very likely wrong. Is it easy to add a test for this (apologies if I missed it), since `struct Foo(x, y)` is like `Foo = \x -> \y -> Foo(x, y)`

Which, by the way, do we have eta conversion for constructor functions? Seems like we need that too.

Reply:
yeah... that needs to be checked. I keep straining to wrap my head around which way the indexes go that which probably means it needs to be refactored (but at minimum tested)

One way to think about it is that the constructor uses the outer most Lambda first (ie the inner most App applies to the outer most Lambda). So the first argument of the struct will have the highest `LambdaVar` index... I think?

Note that constructor functions are just `Lambda`s that return a `Struct`. So beta reduction should handle that. But I need tests (and to actually use the beta reduction). I realized the other day from something you said that we could eliminate `Struct` and just have special constructor function expression. The latter feels a little more correct, but it feels harder to implement with taking advantage of reductions of `Match`. I'll go forward this way, get the tests working, then we can refactor with tests keeping us safe.